### PR TITLE
Foundation: port NSThread to Windows

### PR DESCRIPTION
**Pull request by compnerd**
Implement the Thread interfaces for Windows.  Extend the CoreFoundation
interface to support thread stack reservation sizing.

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

File: Foundation/Thread.swift [hunk 1]
```diff
@@ -183,6 +229,10 @@ open class Thread : NSObject {
         } else {
             _thread = nil
         }
+#elseif os(Windows)
+        _thread = self.withRetainedReference {
+            return _CFThreadCreate(_attr, NSThreadStart, $0)
+        }
```
Comment (spevans) [inline Foundation/Thread.swift]: Isnt this block redundant as it is a duplicate of the block below?

File: Foundation/Thread.swift [hunk 2]
```diff
@@ -248,10 +312,13 @@ open class Thread : NSObject {
     private class func backtraceAddresses<T>(_ body: (UnsafeMutablePointer<UnsafeMutableRawPointer?>, Int) -> [T]) -> [T] {
         // Same as swift/stdlib/public/runtime/Errors.cpp backtrace
         let maxSupportedStackDepth = 128;
-        let addrs = UnsafeMutablePointer<UnsafeMutableRawPointer?>.allocate(capacity: maxSupportedStackDepth)
+        var addrs = UnsafeMutablePointer<UnsafeMutableRawPointer?>.allocate(capacity: maxSupportedStackDepth)
```
Comment (spevans) [inline Foundation/Thread.swift]: `addrs` doesnt seem to be mutated anywhere so can this stay as a `let`?

File: Foundation/Thread.swift (hunk 1 above)
Comment (compnerd) [reply Foundation/Thread.swift]: Not when it was originally written :-).

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@jrose-apple - suggestions as to what may be the cause of the deserialization issue?

**Comment by jrose-apple** [comment]
Does SwiftFoundation on Mac use the system CoreFoundation?

**Comment by phausler** [comment]
It better not be using the system CF. The layout of the objects are different (and we have functions with differing signatures etc)

**Comment by jrose-apple** [comment]
Hm, good point. We probably would have noticed that a long time ago.

Maybe the CoreFoundation overlay is causing problems? Do we need a SwiftCoreFoundation?

**Comment by compnerd** [comment]
@phausler agreed ... that would be really bad, right, I didn't consider that (sorry, dealing with Linux, Android, and Windows a lot atm, so keep forgetting that Darwin has CF there)

**Comment by jrose-apple** [comment]
Or is _CFThreadRef only defined when certain macros are passed? If that's the case, it can't be used in a stored property (yet).

**Comment by phausler** [comment]
_CFThreadRef is only defined in the SCLF version of CF. The Darwin CF does not have such a notion.(also that typedef is private anyhow)

**Comment by compnerd** [comment]
Woah, I think @phausler is on to something ... I think that it is using the **system** CF.  This is not in the Foundation build, but in XCTest.

**Comment by compnerd** [comment]
Hmm, could it be a difference between the Xcode and the CMake build?  It seems from the jenkins build that this is being invoked from xcodebuild.

File: Foundation/Thread.swift [hunk 3]
```diff
@@ -130,37 +165,48 @@ open class Thread : NSObject {
...
 #endif
     internal var _status = _NSThreadStatus.initialized
     internal var _cancelled = false
-    
+
     open private(set) var threadDictionary: NSMutableDictionary = NSMutableDictionary()
 
-    internal init(thread: pthread_t) {
-        // Note: even on Darwin this is a non-optional pthread_t; this is only used for valid threads, which are never null pointers.
+    internal init(thread: _CFThreadRef) {
```
Comment (phausler) [inline Foundation/Thread.swift]: this might be the cause of the failure: because it is marked as internal?

Comment (compnerd) [reply Foundation/Thread.swift]: Well, it was previously `internal` ... we should be allowed to have internal interfaces that are unexposed right?

Comment (phausler) [reply Foundation/Thread.swift]: it was previously the type of pthread_t which was from the Darwin module, where as _CFThreadRef pulls in the CoreFoundation module which has a naming conflict with the system CoreFoundation from my limited understanding.

Comment (phausler) [reply Foundation/Thread.swift]: specifically I think this might be associated with things around inlining visibility.

**Comment by phausler** [comment]
> Hm, good point. We probably would have noticed that a long time ago.
> 
> Maybe the CoreFoundation overlay is causing problems? Do we need a SwiftCoreFoundation?

renaming it to SwiftCoreFoundation has other issues that crop up. We tried that early on in the initial bring-up of SCLF. It failed pretty horribly then, but perhaps with alot of work it could be done now that things are not so much in flux as they were back then?

File: Foundation/Thread.swift (hunk 3 above)
Comment (compnerd) [reply Foundation/Thread.swift]: Hmm, I see how that got away in the past.  Suggestions on how to possibly address this then?

Comment (phausler) [reply Foundation/Thread.swift]: there are two potential solutions here:

1) #ifdef the windows version to use the underlying type that _CFThreadRef is typedef'd from and leave the darwin version to be pthread_t
2) make this method fileprivate if possible.

Comment (jrose-apple) [reply Foundation/Thread.swift]: Another hackaround: you could move the typedef into the Swift module, and then check that it's the same type as _CFThreadRef.

Comment (phausler) [reply Foundation/Thread.swift]: it seems that importing `typedef` is acting more like `newtype` than `typealias`.

Comment (jrose-apple) [reply Foundation/Thread.swift]: I *hope* it's not being treated as a CF type. That would be bad. But I *think* it's still a typealias; the compiler is just trying hard to preserve the sugar even though it doesn't need to.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

File: Foundation/Thread.swift (hunk 3 above)
Comment (compnerd) [reply Foundation/Thread.swift]: Okay, making it `fileprivate` is insufficient, its still failing on the import.  Does it make sense to actually `#if` check and desugar the conditional for Darwin and add a SR for ir?  Or just skip the SR entirely?

Comment (compnerd) [reply Foundation/Thread.swift]: @jrose-apple - can't just move the typedef as there are references to it in the C code.  I guess I could create a local typealias with a double underscore and see if that can get converted on the way down.

Comment (jrose-apple) [reply Foundation/Thread.swift]: You need to make something that *doesn't* reference the C one but is equivalent to it. Which stinks, but is probably the least bad answer.

Comment (compnerd) [reply Foundation/Thread.swift]: @jrose-apple - okay; think that this deserves a SR?

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

File: Foundation/Thread.swift (hunk 3 above)
Comment (compnerd) [reply Foundation/Thread.swift]: Hmm, this is something that is odd, so I've created SR-9811.

**Comment by compnerd** [comment]
@swift-ci please test and merge

**Pull request merged**